### PR TITLE
perf(apps): Extract app code from `archive` at install

### LINKED ISSUE #10308: `getPluginRows` postgres query needs optimization
## Bug description

Heroku is reporting that the following query is super slow:

image

We should fix this

## Environment

- [x] PostHog Cloud
- [ ] self-hosted PostHog (ClickHouse-based), version/commit: _please provide_
- [ ] self-hosted PostHog (Postgres-based, legacy), version/commit: _please provide_

## Additional context

cc @Twixes interested in grabbing this?

### PR DESCRIPTION
**Pull request by Twixes**
## Problem

From #10308:

> I digged into this a bit and the main culprit seems to be… IO: the `getPluginRows()` query returns ~45 MB of data in the form of the `posthog_plugin.archive` blob column. With `archive` excluded, it completes in a reasonable 200-300 ms, but `*` includes `archive`, balooning that. One way to improve this would be to save all files needed to run a plugin into `posthog_pluginsourcefile` at plugin installation time – there are only three we need: `plugin.json`, `index.ts` (or an alias defined in `plugin.json`'s `main`) and now `frontend.tsx`. Most of archive contents is junk at runtime, if we only loaded the required code this wouldn't be an issue (in the medium term at least).

## Changes

This does exactly what I described above.`posthog_plugin.archive` is no longer fetched at all in the plugin server, the necessary source code instead comes from `posthog_pluginsourcefile`, so that we don't have to waste IO on junk data.
Also allows us to remove unpacking utils from the plugin server.

#### EDIT:

For a safer migration process, this only contains Django-side changes. This way if the migration is unsuccessful due to an unforeseen issue, the plugin server is not impacted at all. If there's no trouble though, the plugin server will switch to this mechanism with #10374.

## How did you test this code?

~~WIP, marked as draft as I add test coverage.~~ Added a bunch of tests.

### CODE REVIEW

File: posthog/migrations/0243_unpack_plugin_source_files.py [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+            # The original name of the file is not preserved, but this greatly simplifies the rest of the code,
+            # and we don't need to model the whole filesystem (at this point)
+            PluginSourceFile.objects.create(plugin=plugin, filename="index.ts", source=index_ts)
+
+    # Source plugins have already been migrated in 0233_plugin_source_file, while local ones don't store code in the DB
+    for plugin in Plugin.objects.exclude(plugin_type="source").exclude(plugin_type="local"):
+        try:
+            update_or_create_from_plugin_archive(plugin)
+        except exceptions.ValidationError as e:
+            logger.warn(f"Migration 0243 - failed to extract or save code of plugin {plugin.name} ID {plugin.id}: {e}")
```
Comment (macobo) [inline posthog/migrations/0243_unpack_plugin_source_files.py:68]: Unclear: Is this actionable somehow? What is the user supposed to do.

File: posthog/migrations/0243_unpack_plugin_source_files.py [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+        else:
+            logger.info(f"Migration 0243 - extracted and saved code of plugin {plugin.name} ID {plugin.id}")
```
Comment (macobo) [inline posthog/migrations/0243_unpack_plugin_source_files.py:70]: - This should be `debug`
- You're abusing the logger - never interpolate, always pass interpolated arguments after the logging string. This allows us to leverage better log processing tools once we're off of ECS. (Same goes for other logging calls.

File: posthog/models/plugin.py [hunk 3]
```diff
@@ -232,6 +241,57 @@ class PluginLogEntryType(str, Enum):
...
+            plugin_json = cast(Optional[Dict[str, Any]], get_file_from_archive(plugin.archive, "plugin.json"))
+            if not plugin_json:
+                raise exceptions.ValidationError(f"Could not find plugin.json in plugin {plugin.name} ID {plugin.id}")
+        # Extract frontend.tsx - optional
+        frontend_tsx: Optional[str] = get_file_from_archive(plugin.archive, "frontend.tsx", json_parse=False)
+        # Extract index.ts - optional if frontend.tsx is present, otherwise required
+        main_filename_defined = plugin_json.get("main")
+        main_filenames_to_try = [main_filename_defined] if main_filename_defined else ["index.js", "index.ts"]
+        index_ts: Optional[str] = None
+        for main_filename in main_filenames_to_try:
```
Comment (macobo) [inline posthog/models/plugin.py:266]: Suggestion: Extract function - this is hard to follow for no real reason.

**Comment by macobo** [comment]
I'm not sure I am the right person to review this from a correctness standpoint. Should we get @mariusandra or @yakkomajuri involved?

File: posthog/migrations/0243_unpack_plugin_source_files.py [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+
+
+def reverse_func(apps, schema_editor):
+    logger.info("Migration 0243 - revert started")
+    PluginSourceFile = apps.get_model("posthog", "PluginSourceFile")
+    PluginSourceFile.objects.filter(plugin__plugin_type__in=["source", "local"]).delete()
+    logger.info("Migration 0243 - revert finished")
+
+
+class Migration(migrations.Migration):
```
Comment (macobo) [inline posthog/migrations/0243_unpack_plugin_source_files.py:82]: Should we make this non-atomic? If this can take a while, this reduces operational risk at the cost of consistency.

File: posthog/migrations/0243_unpack_plugin_source_files.py [hunk 5]
```diff
@@ -65,9 +65,11 @@ def update_or_create_from_plugin_archive(plugin):
         try:
             update_or_create_from_plugin_archive(plugin)
         except exceptions.ValidationError as e:
-            logger.warn(f"Migration 0243 - failed to extract or save code of plugin {plugin.name} ID {plugin.id}: {e}")
+            logger.warn(
+                f"Migration 0243 - skipping plugin {plugin.name} ID {plugin.id}, failed to extract or save its code: {e}"
+            )
         else:
-            logger.info(f"Migration 0243 - extracted and saved code of plugin {plugin.name} ID {plugin.id}")
+            logger.debug(f"Migration 0243 - extracted and saved code of plugin {plugin.name} ID {plugin.id}")
```
Comment (macobo) [inline posthog/migrations/0243_unpack_plugin_source_files.py:72]: Avoid string interpolation when logging.

```suggestion
            logger.debug("Migration 0243 - extracted and saved code of plugin." plugin=plugin.name, plugin_id=plugin.id)
```

File: posthog/migrations/0243_unpack_plugin_source_files.py [hunk 6]
```diff
@@ -65,9 +65,11 @@ def update_or_create_from_plugin_archive(plugin):
         try:
             update_or_create_from_plugin_archive(plugin)
         except exceptions.ValidationError as e:
-            logger.warn(f"Migration 0243 - failed to extract or save code of plugin {plugin.name} ID {plugin.id}: {e}")
+            logger.warn(
+                f"Migration 0243 - skipping plugin {plugin.name} ID {plugin.id}, failed to extract or save its code: {e}"
```
Comment (macobo) [inline posthog/migrations/0243_unpack_plugin_source_files.py:69]:
```suggestion
                f"Migration 0243 - skipping plugin, failed to extract or save its code.", plugin=plugin.name, plugin_id=plugin.id, error=e
```

**Comment by Twixes** [comment]
I would indeed appreciate Marius's input here, however he's on holiday right now (and he did approve of the Django-side extraction idea). I guess I can loop @yakkomajuri in for review too.

File: posthog/plugins/utils.py [hunk 7]
```diff
@@ -249,6 +249,46 @@ def get_file_from_archive(archive: bytes, filename: str, *, json_parse: bool = T
         return None
 
 
+def find_index_ts_in_archive(archive: bytes, main_filename: Optional[str] = None) -> str:
+    main_filenames_to_try = [main_filename] if main_filename else ["index.js", "index.ts"]
+    for main_filename in main_filenames_to_try:
+        if index_ts := get_file_from_archive(archive, main_filename, json_parse=False):
```
Comment (macobo) [inline posthog/plugins/utils.py:255]: Python style nit: explicit is better than implicit:

```
index_ts = get_file_from_archive(archive, main_filename, json_parse=False)
if index_ts is not None: return index_ts
```

File: posthog/plugins/utils.py [hunk 8]
```diff
@@ -249,6 +249,46 @@ def get_file_from_archive(archive: bytes, filename: str, *, json_parse: bool = T
...
+    archive: bytes, plugin_json_parsed: Optional[Dict[str, Any]] = None
+) -> Tuple[str, Optional[str], Optional[str]]:
+    """Extract plugin.json, index.ts (which can be aliased) and frontend.tsx out of an archive.
+
+        If plugin.json has already been parsed before this is called, its value can be passed in as an optimization."""
+    if archive is None:
+        raise ValueError(f"There is no archive to extract code from")
+    # Extract plugin.json - required, might be provided already
+    plugin_json: str
+    if not plugin_json_parsed:
```
Comment (macobo) [inline posthog/plugins/utils.py:270]: So there are two cases here: None and empty dictionary. are both of these supposed to behave identically?

File: posthog/migrations/0243_unpack_plugin_source_files.py (hunk 1 above)
Comment (Twixes) [reply posthog/migrations/0243_unpack_plugin_source_files.py]: This is just a warning – in some unlikely case the archive might be broken (though the user should know about this already, as such a plugin could not be enabled). Used "skipping" wording to indicate that this is not an error that MUST be taken care of.

File: posthog/plugins/utils.py (hunk 8 above)
Comment (Twixes) [reply posthog/plugins/utils.py:270]: Yes, at least `name` needs to be in the object

File: posthog/migrations/0243_unpack_plugin_source_files.py [hunk 9]
```diff
@@ -0,0 +1,92 @@
...
+        else:
+            logger.debug("Migration 0243 - extracted and saved code of plugin." plugin=plugin.name, plugin_id=plugin.id)
+
+    logger.info("Migration 0243 - finished")
+
+
+def reverse_func(apps, schema_editor):
+    logger.info("Migration 0243 - revert started")
+    PluginSourceFile = apps.get_model("posthog", "PluginSourceFile")
+    PluginSourceFile.objects.filter(plugin__plugin_type__in=["source", "local"]).delete()
```
Comment (yakkomajuri) [inline posthog/migrations/0243_unpack_plugin_source_files.py:85]: won't this delete rows that were there before this migration?

File: posthog/plugins/utils.py (hunk 8 above)
Comment (Twixes) [reply posthog/plugins/utils.py]: However, in this case it's probably better to only care about the `None` case and warn about `{}` plugin-server-side, so good catch!

File: posthog/migrations/0243_unpack_plugin_source_files.py [hunk 10]
```diff
@@ -0,0 +1,92 @@
...
+        if not plugin_json:
+            raise exceptions.ValidationError(f"Could not find plugin.json in plugin {plugin.name} ID {plugin.id}")
+        # Extract frontend.tsx - optional
+        frontend_tsx: Optional[str] = get_file_from_archive(plugin.archive, "frontend.tsx", json_parse=False)
+        # Extract index.ts - optional if frontend.tsx is present, otherwise required
+        main_filename_defined = plugin_json.get("main")
+        main_filenames_to_try = [main_filename_defined] if main_filename_defined else ["index.js", "index.ts"]
+        index_ts: Optional[str] = None
+        for main_filename in main_filenames_to_try:
+            if index_ts := get_file_from_archive(plugin.archive, main_filename, json_parse=False):
```
Comment (yakkomajuri) [inline posthog/migrations/0243_unpack_plugin_source_files.py:45]: unsure what `get_file_from_archive` returns, but what if my `index.ts` is an empty string? We should probably create a source file from it for compatibility?

File: posthog/migrations/0243_unpack_plugin_source_files.py [hunk 11]
```diff
@@ -0,0 +1,92 @@
...
+                    f"Could not find main file {' or '.join(main_filenames_to_try)} in plugin {plugin.name} ID {plugin.id}"
+                )
+        # Save plugin.json
+        PluginSourceFile.objects.create(plugin=plugin, filename="plugin.json", source=json.dumps(plugin_json))
+        # Save frontend.tsx
+        if frontend_tsx:
+            PluginSourceFile.objects.create(plugin=plugin, filename="frontend.tsx", source=frontend_tsx)
+        # Save index.ts
+        if index_ts:
+            # The original name of the file is not preserved, but this greatly simplifies the rest of the code,
```
Comment (yakkomajuri) [inline posthog/migrations/0243_unpack_plugin_source_files.py:59]: as long as this doesn't cause problems elsewhere? haven't read the API changes yet so maybe that will answer my question - but what if someone then updates the source of their `main.js` file?

File: posthog/migrations/0243_unpack_plugin_source_files.py (hunk 1 above)
Comment (macobo) [reply posthog/migrations/0243_unpack_plugin_source_files.py]: I can see a user upgrading and getting confused or asking about this warning. Do you think we should reword to minimize that pain?

**Review by yakkomajuri — COMMENTED** [review]
thanks for all the comments in the code.

looking good, left a few questions for now

File: posthog/migrations/0243_unpack_plugin_source_files.py (hunk 9 above)
Comment (Twixes) [reply posthog/migrations/0243_unpack_plugin_source_files.py:85]: Oops, of course, this should be an `.exclude()` instead of a `.filter()`

File: posthog/migrations/0243_unpack_plugin_source_files.py (hunk 10 above)
Comment (Twixes) [reply posthog/migrations/0243_unpack_plugin_source_files.py:45]: Fair point, so the walrus is gone

File: posthog/migrations/0243_unpack_plugin_source_files.py (hunk 1 above)
Comment (Twixes) [reply posthog/migrations/0243_unpack_plugin_source_files.py]: This should never really happen, so I think the message is fine after the update – the user probably _should_ message us if they see it, as this either means their custom plugin stored in GitHub/npm is malformed (and if they don't know about this, they'll have a bad time anyway) or an approved repository plugin is malformed (but I just ran the migration with the majority of repository plugins and found no problems).

File: posthog/migrations/0243_unpack_plugin_source_files.py (hunk 4 above)
Comment (yakkomajuri) [reply posthog/migrations/0243_unpack_plugin_source_files.py:69]: This is fine 🔥

**Review by yakkomajuri — APPROVED** [review]
Ultimately the person with the most context here would be @mariusandra so if we want to be super sure we can loop him in.

From my perspective, my concerns were addressed so this LGTM. Left one small comment.

File: posthog/test/test_plugin.py [hunk 12]
```diff
@@ -20,3 +37,155 @@ def test_default_config_dict(self):
...
+        self.assertIsNone(frontend_tsx_file)
+
+    def test_sync_from_plugin_archive_twice_from_zip_with_index_ts_works(self):
+        test_plugin: Plugin = Plugin.objects.create(
+            organization=self.organization,
+            name="Contoso",
+            archive=base64.b64decode(HELLO_WORLD_PLUGIN_RAW_WITH_INDEX_TS_BUT_UNDEFINED_MAIN),
+        )
+
+        with self.assertNumQueries(13):
```
[the remark this answers is not in the document]
Comment (macobo) [reply posthog/test/test_plugin.py:119]: A better solution is @snapshot_postgres_queries

[the remark this answers is not in the document]
Comment (Twixes) [reply posthog/test/test_plugin.py]: I was looking for exactly this, thanks @macobo!
